src/plugins/fs/vfat.c

```diff
@@ -328,6 +325,60 @@ gboolean bd_fs_vfat_check_label (const gchar *label, GError **error) {
     return TRUE;
 }
 
+/**
+ * bd_fs_vfat_set_uuid:
+ * @device: the device containing the file system to set uuid for
+ * @uuid: (nullable): volume ID to set or %NULL to generate a new one
+ * @error: (out) (optional): place to store error (if any)
+ *
+ * Returns: whether the volume ID of vfat file system on the @device was
+ *          successfully set or not
+ *
+ * Tech category: %BD_FS_TECH_VFAT-%BD_FS_TECH_MODE_SET_UUID
+ */
+gboolean bd_fs_vfat_set_uuid (const gchar *device, const gchar *uuid, GError **error) {
+    const gchar *args[5] = {"fatlabel", "-i", device, NULL, NULL};
+
+    if (!check_deps (&avail_deps, DEPS_FATLABELUUID_MASK, deps, DEPS_LAST, &deps_check_lock, error))
+        return FALSE;
+
+    if (!uuid || g_strcmp0 (uuid, "") == 0)
+        args[3] = "--reset";
+    else
+        args[3] = uuid;
+
+    return bd_utils_exec_and_report_error (args, NULL, error);
+}
+
+/**
+ * bd_fs_vfat_check_uuid:
+ * @uuid: UUID to check
+ * @error: (out) (optional): place to store error
+ *
+ * Returns: whether @uuid is a valid UUID for the vfat file system or not
+ *          (reason is provided in @error)
+ *
+ * Tech category: always available
+ */
+gboolean bd_fs_vfat_check_uuid (const gchar *uuid, GError **error) {
+    guint64 vol_id;
+
+    vol_id = g_ascii_strtoull (uuid, NULL, 16);
```

Comment:
I was also wondering whether we should accept VOLID strings as presented by udev - i.e. not just a plain 32-bit hex number, but also the `XXXX-YYYY` form. I would certainly like to be this way for UDisks since we present it that form and thinking about the right place to reformat the ID.

This needs to be consistent across all filesystem plugins. Most of them accept everything `uuid_parse()` accepts such as `%08x-%04x-%04x-%04x-%012x`, however e.g. NTFS also needs hex string without dashes.

Reply:
Another problem here is that `g_ascii_strtoull()` can stop conversion halfways, returning partially converted value. `endptr` needs to be checked (the second argument) and proper error returned.

For example: input string `28AA-7D6D` gets converted to `28AA` and resulting VOLID becomes `0000-28AA`.

Reply:
> I was also wondering whether we should accept VOLID strings as presented by udev - i.e. not just a plain 32-bit hex number, but also the `XXXX-YYYY` form. I would certainly like to be this way for UDisks since we present it that form and thinking about the right place to reformat the ID.

Ok, I changed the code to accept vol IDs with the dash in the middle.

> 
> This needs to be consistent across all filesystem plugins. Most of them accept everything `uuid_parse()` accepts such as `%08x-%04x-%04x-%04x-%012x`, however e.g. NTFS also needs hex string without dashes.

Yes, but for NTFS you get the ID from blkid/udev without dashes, e.g. `4A514A2C13B7198B`.

> Another problem here is that `g_ascii_strtoull()` can stop conversion halfways, returning partially converted value. `endptr` needs to be checked (the second argument) and proper error returned.
> 
> For example: input string `28AA-7D6D` gets converted to `28AA` and resulting VOLID becomes `0000-28AA`.

Oh, I didn't realize that, I thought it will fail to parse it. It should be fixed now.

Reply:
At the end I've made udisks to remove any dashes for certain filesystem types, so we should be covered from both sides.